pkg/services/sqlstore/migrations/ualert/alert_rule.go

```diff
@@ -206,6 +207,56 @@ func fixGraphiteReferencedSubQueries(queryData map[string]json.RawMessage) map[s
 	return queryData
 }
 
+// fixPrometheusBothTypeQuery converts Prometheus 'Both' type queries to range queries.
+func fixPrometheusBothTypeQuery(l log.Logger, queryData map[string]json.RawMessage) map[string]json.RawMessage {
+	// There is the possibility to support this functionality by:
+	//	- Splitting the query into two: one for instant and one for range.
+	//  - Splitting the condition into two: one for each query, separated by OR.
+	// However, relying on a 'Both' query instead of multiple conditions to do this in legacy is likely
+	// to be unintentional. In addition, this would require more robust operator precedence in classic conditions.
+	// Given these reasons, we opt to convert them to range queries and log a warning.
+
+	// Check if the query is for Prometheus.
+	ds, ok := queryData["datasource"]
+	if !ok {
+		return queryData
+	}
+	var datasource struct {
+		Type string `json:"type"`
+	}
+	if err := json.Unmarshal(ds, &datasource); err != nil {
+		// Nothing to do here, we can't parse the datasource.
+		return queryData
+	}
+	if datasource.Type != "prometheus" {
+		// Only apply this fix to Prometheus.
+		return queryData
+	}
+
+	// Convert 'Both' type queries to range queries.
+	var instant bool
+	if instantRaw, ok := queryData["instant"]; ok {
+		if err := json.Unmarshal(instantRaw, &instant); err != nil {
+			// Nothing to do here, we can't parse the instant field.
+			return queryData
+		}
+	}
+	var rng bool
+	if rangeRaw, ok := queryData["range"]; ok {
+		if err := json.Unmarshal(rangeRaw, &rng); err != nil {
+			// Nothing to do here, we can't parse the instant field.
+			return queryData
```

Comment:
What do you think about leaving a debug\info message when we fail to parse the payload? I think it would help track down "unsuccessful" migrations. I would do Info level.

Reply:
Makes sense, I can add a log to the unsuccessful unmarshalling of `datasource`, `instant` and `range`.